File: src/odioctl/netinfo.py
"""Host network helpers: LAN-reachable IP and the PWA deep link."""

from __future__ import annotations

import argparse
import re
import subprocess

PWA_URL = "https://pwa.odio.love"


def default_route_ip() -> str | None:
    """Source IP of the default route, or None when it can't be determined
    (no default route, `ip` missing, unparseable output)."""
    try:
        out = subprocess.run(
            ["ip", "route", "get", "1.1.1.1"],
            capture_output=True,
            text=True,
            check=True,
        ).stdout
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None
    m = re.search(r"\bsrc\s+(\S+)", out)
    return m.group(1) if m else None


def pwa_url() -> str:
    """https://pwa.odio.love/#/i/<ip>, or the bare PWA URL when no IP is
    detectable so callers (motd, post-install summary) always get something
    printable."""
    ip = default_route_ip()
    return f"{PWA_URL}/#/i/{ip}" if ip else PWA_URL


def add_pwa_url_arguments(p: argparse.ArgumentParser) -> None:
    p.description = (
        "Print the PWA URL pointing at this host's LAN-reachable IP "
        "(the source IP of the default route)."
    )


def cmd_pwa_url(_ns: argparse.Namespace) -> int:
    print(pwa_url())
    return 0

File: netinfo/netinfo.go
// Package netinfo answers "how does the LAN reach this box": the source IP
// of the default route, and the PWA deep link built from it.
package netinfo

import "net"

const BaseURL = "https://pwa.odio.love"

// DefaultRouteIP is the source IP of the default route, "" when there is
// none. A UDP dial makes the kernel pick route and source address without
// sending a packet.
func DefaultRouteIP() string {
	conn, err := net.Dial("udp4", "1.1.1.1:53")
	if err != nil {
		return ""
	}
	defer func() { _ = conn.Close() }()
	if addr, ok := conn.LocalAddr().(*net.UDPAddr); ok {
		return addr.IP.String()
	}
	return ""
}

// PWAURL is https://pwa.odio.love/#/i/<ip>, or the bare PWA URL when no IP is
// detectable, so callers (motd, post-install summary) always get something
// printable.
func PWAURL() string {
	if ip := DefaultRouteIP(); ip != "" {
		return BaseURL + "/#/i/" + ip
	}
	return BaseURL
}
